source/extensions/common/wasm/wasm.cc

```diff
@@ -165,45 +160,87 @@ static void createWasmInternal(const VmConfig& vm_config, const PluginSharedPtr&
   std::string source, code;
+  bool fetch = false;
   if (vm_config.code().has_remote()) {
     source = vm_config.code().remote().http_uri().uri();
+    std::lock_guard<std::mutex> guard(code_cache_mutex);
+    if (!code_cache) {
+      code_cache = new std::remove_reference<decltype(*code_cache)>::type;
+    }
+    auto it = code_cache->find(vm_config.code().remote().sha256());
+    if (it != code_cache->end()) {
+      if (it->second.in_progress) {
+        throw WasmException(
+            fmt::format("Failed to load WASM code (fetch in progress) from {}", source));
+      }
+      code = it->second.code;
+      if (code.empty()) {
+        fetch = true; // Fetch failed, retry.
+        it->second.in_progress = true;
+      }
+    } else {
+      fetch = true; // Not in cache, fetch.
+      (*code_cache)[vm_config.code().remote().sha256()].in_progress = true;
+    }
   } else if (vm_config.code().has_local()) {
     code = Config::DataSource::read(vm_config.code().local(), true, api);
     source = Config::DataSource::getPath(vm_config.code().local())
                  .value_or(code.empty() ? EMPTY_STRING : INLINE_STRING);
   }
 
-  auto callback = [vm_config, scope, &cluster_manager, &dispatcher, &lifecycle_notifier, plugin, cb,
-                   source,
-                   context_ptr = root_context_for_testing ? root_context_for_testing.release()
-                                                          : nullptr](const std::string& code) {
-    if (code.empty()) {
-      throw WasmException(fmt::format("Failed to load WASM code from {}", source));
-    }
-    std::unique_ptr<Context> context(context_ptr);
-    auto configuration = vm_config.configuration();
-    auto vm_key = proxy_wasm::makeVmKey(vm_config.vm_id(), configuration, code);
-    cb(std::static_pointer_cast<WasmHandle>(proxy_wasm::createWasm(
-        vm_key, code, plugin,
-        [&vm_config, &configuration, &scope, &cluster_manager, &dispatcher,
-         &lifecycle_notifier](absl::string_view vm_key) {
-          auto wasm = std::make_shared<Wasm>(vm_config.runtime(), vm_config.vm_id(), configuration,
-                                             vm_key, scope, cluster_manager, dispatcher);
-          // NB: we need the shared_ptr to have been created for shared_from_this() to work.
-          wasm->initializeLifecycle(lifecycle_notifier);
-          return std::static_pointer_cast<WasmHandleBase>(std::make_shared<WasmHandle>(wasm));
-        },
-        vm_config.allow_precompiled(), std::move(context))));
-  };
+  auto complete_cb =
+      [cb, vm_config, plugin, scope, &cluster_manager, &dispatcher, &lifecycle_notifier,
+       root_context_for_testing_ptr = root_context_for_testing.release()](std::string code) {
+        auto root_context = std::unique_ptr<Context>(root_context_for_testing_ptr);
+        auto vm_key = proxy_wasm::makeVmKey(vm_config.vm_id(), vm_config.configuration(), code);
+        cb(std::static_pointer_cast<WasmHandle>(proxy_wasm::createWasm(
```

Comment:
can we add some metrics here?
How many fetches, retries? Unless it is already tracked in cluster metrics.
Perhaps it would still be better to track those here.

Reply:
The retries are hidden by the RemoteAsyncDataProvider.  I can tell you how many fetches there are, but as you say, that is covered by the cluster stats.  However, if you want to know how many fetches there were and the disposition thereof I can add that.   I would imagine though that given how infrequently we update the .wasm objects those stats are not going to be very interesting.   It will only be interesting if the cluster is down.

Reply:
Think about it and tell me what you want.  I can add it here or in a follow up PR.